### PR TITLE
Add API to add artifacts

### PR DESCRIPTION
**Pull request by aguschin**
WIP
- [ ] ~~implement `dvc get` artifacts~~
- [x] add writing new artifacts (used by Studio, DVCLive uses its own way due to simplicity)
- [x] Remove some name limitations: "Do you know if using a number as the first character is problematic? If not, could we drop the restriction for The first character must be a letter.? The rest of the logic seems pretty straightforward."
- [ ] ~~assert that `get` is implemented right~~
  - [ ] ~~assert it's alright to simply split by `:` to get `dvc.yaml` location and artifact name~~
- [ ] ~~implement same logic for `import`~~
- [x] add tests

### CODE REVIEW

**Comment by dberenbaum** [comment]
@aguschin Converted to draft since it's WIP. Hope that's okay.

File: dvc/dvcfile.py [hunk 1]
```diff
@@ -289,6 +289,33 @@ def _dump_pipeline_file(self, stage):
 
         self.repo.scm_context.track_file(self.relpath)
 
+    def _dump_pipeline_file_artifacts(self, name, artifact):
+        # I can either generalize `_dump_pipeline_file` or implement a separate one
+        # happens that `_dump_pipeline_file` is very much tailored to the `Stage` class
+        # WDYT - does it worth to generalize?
```
Comment (aguschin) [inline dvc/dvcfile.py:295]: This looks pretty ugly to me. Let's make a conscious decision here: either we make this OK here, or I simply move this logic to Studio. I would really like to keep it in DVC since it's both easier to use and change. But if that's will be very tough to do and it won't worth it, moving it to Studio is fine. @skshetry can you please help me understand how this can be done better if we keep it in DVC?

Comment (omesser) [reply dvc/dvcfile.py:295]: > I would really like to keep it in DVC
➕

File: dvc/repo/get.py [hunk 2]
```diff
@@ -20,13 +21,27 @@ def __init__(self):
...
+    parts = path.split(":")
+    if len(parts) == 1:
+        dvcyaml = None
+    elif len(parts) == 2:
+        dvcyaml, path = parts
+    else:
+        raise ValueError("more than one ':' found")  # todo replace error
+    if dvcyaml is not None and dvcyaml[-len(PROJECT_FILE) :] != PROJECT_FILE:
+        dvcyaml = os.path.join(dvcyaml, PROJECT_FILE)
+    return dvcyaml, path
```
[the remark this answers is not in the document]
Comment (omesser) [reply dvc/repo/get.py:34]: IMO A robust approach seems to be this: if ":" exists, split by the first occurrence of it (if there are more, will be part of the `path` since it's last

File: dvc/repo/get.py [hunk 3]
```diff
@@ -20,13 +21,27 @@ def __init__(self):
         )
 
 
+def resolve_artifact_addressing(path):  # add tests
```
Comment (omesser) [inline dvc/repo/get.py:24]:
```suggestion
def resolve_artifact_path(path):
```
Is this always a file path? "path" is consistent with the variable name returned
if it's more generic it should be "url" or "address" I think - consistently though 😄

File: dvc/repo/get.py [hunk 4]
```diff
@@ -61,6 +76,21 @@ def get(url, path, out=None, rev=None, jobs=None):
...
 
+            if dvcyaml:
+                artifacts = repo.artifacts.read()
+                if dvcyaml not in artifacts:
+                    raise InvalidArgumentError(
+                        f"{dvcyaml} doesn't exist or doesn't have artifacts section"
+                    )
+                if path not in artifacts[dvcyaml]:
+                    raise InvalidArgumentError(
+                        f"artifact {path} doesn't exist in {dvcyaml}"
```
Comment (omesser) [inline dvc/repo/get.py:87]:
```suggestion
                        f"Artifact '{path}' doesn't exist in '{dvcyaml}'"
```
😅

File: dvc/repo/get.py [hunk 5]
```diff
@@ -61,6 +76,21 @@ def get(url, path, out=None, rev=None, jobs=None):
         ) as repo:
             from dvc.fs.data import DataFileSystem
 
+            if dvcyaml:
+                artifacts = repo.artifacts.read()
+                if dvcyaml not in artifacts:
+                    raise InvalidArgumentError(
+                        f"{dvcyaml} doesn't exist or doesn't have artifacts section"
```
Comment (omesser) [inline dvc/repo/get.py:83]:
```suggestion
                        f"'{dvcyaml}' doesn't exist or doesn't have an artifacts section"
```

File: dvc/repo/get.py [hunk 6]
```diff
@@ -61,6 +76,21 @@ def get(url, path, out=None, rev=None, jobs=None):
...
+                    )
+                if path not in artifacts[dvcyaml]:
+                    raise InvalidArgumentError(
+                        f"artifact {path} doesn't exist in {dvcyaml}"
+                    )
+                path = repo.artifacts.read()[dvcyaml][
+                    path
+                ].path  # check keys exist and raise errors if not
+                if path is None:
+                    raise InvalidArgumentError("path is None")  # todo replace error
```
Comment (omesser) [inline dvc/repo/get.py:93]: more meaningful if we include the artifact and dvcyaml

**Review by omesser — CHANGES_REQUESTED** [review]

File: dvc/commands/get.py [hunk 7]
```diff
@@ -61,7 +61,7 @@ def add_parser(subparsers, parent_parser):
         "url", help="Location of DVC or Git repository to download from"
     )
     get_parser.add_argument(
-        "path", help="Path to a file or directory within the repository"
+        "path", help="Path to a file, directory or artifact within the repository"
```
[the remark this answers is not in the document]
Comment (aguschin) [reply dvc/commands/get.py:64]: Yes, it corresponds. But then it's not `path`, but artifact. Let me rephrase this `Path to a file, directory, or artifact ID within the repository`

Comment (aguschin) [inline dvc/commands/get.py:64]:
```suggestion
        "path", help="Path to a file or directory, or artifact ID within the repository"
```

File: dvc/repo/get.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aguschin) [reply dvc/repo/get.py]: We don't allow for `:` in artifact name, so it can't happen. Rather, if there is more that a single `:`, we can treat it as path (not an artifact ID). 🤔 if we should add it though.

[the remark this answers is not in the document]
Comment (aguschin) [reply dvc/repo/get.py]: Posting what we discussed yesterday at refinement: it's fine to keep it like this. In future we'll collect feedback and make it work for other cases we can miss now.

File: dvc/dvcfile.py (hunk 1 above)
Comment (skshetry) [reply dvc/dvcfile.py:295]: Better to move it to `Artifacts` for now.
Also, please decouple "removal" of artifacts from update/add.
And let's skip logging messages for now, as dvc cli is not going to use it.

File: dvc/repo/get.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (skshetry) [reply dvc/repo/get.py]: `:` is valid name for files.

**Comment by dberenbaum** [comment]
@aguschin Is this still in draft mode? Is it ready to be reviewed and merged?

**Comment by dberenbaum** [comment]
@aguschin Let's not include in this PR, but what about `dvc get-url` and API methods (`dvc.api.open/read` and `DVCFileSystem`)?

**Comment by skshetry** [comment]
> @aguschin Let's not include in this PR, but what about `dvc get-url` and API methods (`dvc.api.open/read` and `DVCFileSystem`)?

How would DVCFileSystem implement this? On other APIs, it feels too early.

Also, I am not a fan of overloading target here, I'd have preferred it to be unambiguous with `--model` or similar flags. Or, alternatively get rid of `dvc.yaml` path and assume model names are going to be unique, as there's a way to download using the path unambiguously.

To be honest, I am not seeing much value with this, that's not already possible.

**Comment by aguschin** [comment]
> @aguschin Let's not include in this PR, but what about `dvc get-url` and API methods (`dvc.api.open/read` and `DVCFileSystem`)?

I don't think we should implement it there for the first release - don't see clearly the scenario when it's really needed when there is get/import. If users will need that, we can add it later.

> Also, I am not a fan of overloading target here, I'd have preferred it to be unambiguous with --model or similar flags

Thank you for sharing your opinion @skshetry! I understand you here. My thoughts: we use the same to reference stages in dvc.yaml, so I think it's OK to reuse same approach, since we already took it.

> Or, alternatively get rid of dvc.yaml path and assume model names are going to be unique, as there's a way to download using the path unambiguously.

Rn, names must be unique within a selected dvc.yaml, but I assume you meant they should be unique over all dvc.yaml files in DVC repo? This is quite hard to maintain I think, so don't really see this as an option.

> To be honest, I am not seeing much value with this, that's not already possible.

Sorry, didn't get you here. Did you meant this PR, or `dvc get-url` and API methods (`dvc.api.open/read` and `DVCFileSystem`)?

**Comment by dberenbaum** [comment]
@aguschin How will users get the revisions they want? For example, I may always want the latest prod version. I think this is more important than supporting the artifact name.

**Comment by skshetry** [comment]
> Rn, names must be unique within a selected dvc.yaml, but I assume you meant they should be unique over all dvc.yaml files in DVC repo? This is quite hard to maintain I think, so don't really see this as an option.

It's okay to make an assumption that they are going to be unique on a repo-level without enforcing it, as long as they are documented. There's an alternative way, from path to download artifacts.

**Comment by skshetry** [comment]
> Sorry, didn't get you here. Did you meant this PR, or `dvc get-url` and API methods (`dvc.api.open/read` and `DVCFileSystem`)?

I meant `dvc get`. Users can already download using path. We are only making it more confusing by overloading it.

I don't see much difference between either of the following options. Most of the time, the artifact will refer to the same path, they will rarely change.

```console
$ dvc get https://github.com/iterative/example-get-started dvc.yaml:artifact1
$ dvc get https://github.com/iterative/example-get-started path/model
```

I do see in https://github.com/iterative/dvc/issues/9100 that you proposed:
```console
$ dvc get $REPO mymodel@latest
```

Which I do see value in.

**Comment by dberenbaum** [comment]
> $ dvc get $REPO mymodel@latest

@skshetry Agreed that this is looks more valuable, but I think it makes sense to merge support for artifact names and follow up later with gto revision support.

@aguschin Can we have some kind of revision support in `dvc get` for release? I think it's important since otherwise we are still stuck asking users to install and learn gto to deploy.

**Comment by skshetry** [comment]
Still think of either using a flag or at least getting rid of dvc.yaml path. Requiring too many information is counter-productive especially when the difference is already very minimal between path and name. It’s a remote resource that we are talking about, figuring out where the artifact is defined is even harder.

**Comment by aguschin** [comment]
> @aguschin Can we have some kind of revision support in dvc get for release? I think it's important since otherwise we are still stuck asking users to install and learn gto to deploy.

@dberenbaum, if they did stage assignments with Studio, we expose right revision to Studio's `dvc get` command, so no need to use GTO. If they don't use Studio, they did assignments with GTO, so it's ok to expect them to use GTO to get right revisions. Please correct me if I'm missing something.

> It’s a remote resource that we are talking about, figuring out where the artifact is defined is even harder.

This is a really good point, as well as the other one. WDYT @dberenbaum? 

> It's okay to make an assumption that they are going to be unique on a repo-level without enforcing it, as long as they are documented. There's an alternative way, from path to download artifacts.

@skshetry, OK, I get your point. Please help me understand then: what happens if I have same artifact names in multiple dvc.yamls across the repo? Specially, what happens when I call `dvc get $REPO mymodel`? (maybe there are some other unclear cases, please share if you see them).

And to mention, there still be cases, when knowing the path to dvc.yaml will be required. Like when you have a monorepo like https://github.com/aguschin/monorepo-test-no-root - you must supply `s3` or `ssh` there, otherwise DVC won't know it needs to check those dirs for dvc.yamls and artifacts in them.

**Comment by dberenbaum** [comment]
> @dberenbaum, if they did stage assignments with Studio, we expose right revision to Studio's `dvc get` command, so no need to use GTO. If they don't use Studio, they did assignments with GTO, so it's ok to expect them to use GTO to get right revisions. Please correct me if I'm missing something.

They could be doing stage assignments in Studio but deployments via some automated CD process, right? I think this is the happy path, but let me know if you have a different idea.

Ideally, the `dvc get` command they copy can be reused. The artifact name should be static, but the rev will likely change frequently.

> @skshetry, OK, I get your point. Please help me understand then: what happens if I have same artifact names in multiple dvc.yamls across the repo? Specially, what happens when I call `dvc get $REPO mymodel`? (maybe there are some other unclear cases, please share if you see them).

For `dvc get $REPO mymodel` with multiple `mymodel` artifacts across different `dvc.yaml` files, we could either fail or decide on some priority order, and we could still allow `subdir/dvc.yaml:mymodel`/`subdir/mymodel` for when you need to be more specific. We could also do `art:mymodel`/`model:mymodel` instead of `dvc.yaml:mymodel` which would be more meaningful and less like a path if needed.

Overall, I think @skshetry raises a good point that abstracting the path (which is likely static) seems less important than abstracting the revision. WDYT about focusing there instead @aguschin?

**Comment by aguschin** [comment]
Thanks @dberenbaum! I generally agree - my only doubt is that we need to implement it before the release.

> Ideally, the `dvc get` command they copy can be reused. The artifact name should be static, but the rev will likely change frequently

Sure. In Studio they currently they copy a command for a single _version_, not what's assigned to stage. Such command would be helpful for local/remote use, not in CI/CD, but not THAT useful in CI/CD IMHO.

> They could be doing stage assignments in Studio but deployments via some automated CD process, right? I think this is the happy path, but let me know if you have a different idea.

There are generally 2 cases: CI is triggered by a Git tag, or CI is triggered by something else.

If you run CI/CD, it's triggered by a right revision already (I promoted `mymodel` to `prod` -> that creates git tag -> at push git tag triggers CI/CD). So in this kind of CI/CD you don't need to know revision. You simply can run `dvc get . mymodel`.

In a far less frequent (my gut feeling, please correct me if you disagree) case you'd be running CI/CD deployment by schedule or when it was triggered by some external signal (webhook or whatever). In this case, you'll need to know which version is in prod. But CI/CD usually starts with a shallow cloned repo, so there are technical difficulties there we need to resolve, so it doesn't look like running `dvc get . mymodel#prod` is the last obstacle on that path:

`dvc get . mymodel --rev mymodel#prod#13` will fail with `ERROR: failed to get 'model.pkl' from '.' - unknown Git revision 'mymodel@v0.0.1'` since `mymodel#prod#13` is outside of the shallow clone. Increasing the depth of shallow clone doesn't help, that's why I have some special commands run in Gto-action to fetch all git tags https://github.com/iterative/gto-action/blob/v1/entrypoint.sh#L9-L11

(Maybe instead of fetching all git tags like this user can specify remote repo url like `dvc get $REPO_URL mymodel#prod` and that will works 🤔  - I didn't try this)

To hide all this complexity, they better use GTO action to find out right revision or even run `dvc get` inside that action (I need to update it though to accept not `$GITHUB_REF` only, but a direct instruction - like `mymodel#prod` or `mymodel@latest` - that doesn't work now). And that action extends to GH only so far.

TLDR: using `dvc get . mymodel#prod` for CI won't be that frequent (IMHO) and there is another thing(s) users need to understand, so implementing will help, but not that drastically. That's why I don't think we need to rush to implement this before the release.

**Comment by dberenbaum** [comment]
I was thinking of more "niche" scenarios where, for example, a separate engineering team downloads and deploys the ML team's model in a different repo (for example, as part of a daily job). I think it's pretty common for a more "enterprise" scenario, but I get why you think it's less of a priority, and I agree it's okay to drop it from the initial scope.

> If you run CI/CD, it's triggered by a right revision already (I promoted `mymodel` to `prod` -> that creates git tag -> at push git tag triggers CI/CD). So in this kind of CI/CD you don't need to know revision. You simply can run `dvc get . mymodel`.

So if this is the "happy path," even `dvc get` is optional, right? I could just as easily use `dvc pull`, although suggesting `dvc get` makes sense since it's more universal.

Do you see this PR as critical to that happy path? We will still need to explain a lot about how to setup CD to work with the artifact. Should we deprioritize this PR and focus that time on showing useful instructions in Studio and the docs explaining how to trigger the workflow and get the model?

For example, in GitHub, I guess you need something like this in your workflow file?

```yaml
on:
  push:
    tags:
      - 'mymodel#prod#[0-9]+'
...
jobs:
  deploy:
    steps:
      - uses: iterative/setup-dvc@v1
      - name: get-model
         run: |
           dvc get . path/to/mymodel
           ...
```

**Comment by shcheklein** [comment]
(sorry, hard to follow all the converstation) 

Some of my thoughts. I feels important to have an access MLFlow-style (`dvc get $REPO mymodel@latest` _and_ Python api that corresponds to that). Let's think why:

- simplicity, since I have name - it's easier to write, think about names as aliases, etc (vs full path that I need to remember)
- similarity with how other MRs / artifact registries work. Again, mental load of relying on both sometimes - names and path is bigger. It's something new to manage (pathes), etc
- if we keep path as an ID It means I have 3-5 places (dvc.yaml, .dvc, CI/CD, code ..) with that path hardcoded. Again, seems to be a minor things but it leads to frustration, errors, etc. Imagine a case when I do refactoring and moving an artifact. I can't now even write a CI/CD that would work both (not sure it needed?) and I would have to update the workflow file ... breaks the point of having names in the first place
- onboarding ppl into MR - also easier to "sell" here is the name and stage and you can get a model by it (related to the first item in the list)

(I'm not arguing about specific syntax here btw, overloading dvc get or make it more explicit - this is primarily about being able to address model by their names)

**Comment by dberenbaum** [comment]
@shcheklein @aguschin What would be your priority between these, and which do you think are blockers for release?
1. `dvc get` name support
2. `dvc get` tag support
3. Python API support
4. Import and get-url support

IMO having `dvc get` name support without `dvc get` version/tag support feels incomplete, so I would lean towards doing both (doesn't have to be in the same PR) or following up with both after release rather than doing half of it now.

**Comment by shcheklein** [comment]
> IMO having dvc get name support without dvc get version/tag support feels incomplete, so I would lean towards doing both (doesn't have to be in the same PR) or following up with both after release rather than doing half of it now.

sounds good to me. If we agree on the scope we can do it step by step and ship it as it's getting ready.

> Import and get-url support

Never thought about this even. But sounds also reasonable. It feels it can come "for free" if we do names support in `dvc get`. I wonder what @aguschin thinks here.

**Comment by aguschin** [comment]
Thank you @shcheklein @dberenbaum, this is a very helpful discussion!

> So if this is the "happy path," even dvc get is optional, right? I could just as easily use dvc pull, although suggesting dvc get makes sense since it's more universal.

Yes I think. You still need to figure out what artifact was promoted to the stage and which `path` it has, so a simple `dvc pull` may not be enough if you have 1+ models.

> Never thought about this even. But sounds also reasonable. It feels it can come "for free" if we do names support in dvc get. I wonder what @aguschin thinks here.

`import`ing artifacts can be useful. E.g. I can have a separate repo where I consume few models and build an application. I can `import` a model of a separate revision, and have CI/CD creating PRs to update versions once new versions are promoted to `prod` or `dev` (like `dependabot` does it for packages). I may even want to use not a Git revision, but what's in `prod` directly - `import`ing `mymodel@latest` or `mymodel#prod` like we discuss it to work for DVC get.

In terms of implementation, I'm not sure how involved it would be. 

> IMO having `dvc get` name support without `dvc get` version/tag support feels incomplete, so I would lean towards doing both (doesn't have to be in the same PR) or following up with both after release rather than doing half of it now.

Yeah, agree on "incompleteness" feeling here, even though I think it comes from not able to show `dvc get . mymodel@latest` - which may not be the most useful thing (as CI/CD triggered by tags is I believe), but that creates a feeling something is missing in our MR implementation. 

So I guess it comes down to this plan, sorted by priority:
1. Write docs (with CI/CD examples), release it in Studio (which means I need to merge this PR, at least the part where we can add new artifacts to `dvc.yaml`), release new versions of gto and gto-action.
2. Support `dvc get . mymodel`, `dvc get . mymodel#prod`, Python API support, update docs
3. Support `dvc import`, `dvc get-url` and etc (is there is something else), update docs

So I'm going to de-prioritize `dvc get` then, change this PR to introduce adding new artifacts only, and start drafting docs so we have at least a rough version of those.

File: dvc/repo/artifacts.py [hunk 8]
```diff
@@ -65,3 +70,37 @@ def read(self) -> Dict[str, Dict[str, Artifact]]:
                 check_name_format(name)
                 artifacts[dvcyaml][name] = Artifact(**value)
         return artifacts
+
+    def add(self, name: str, artifact: Artifact, dvcfile: Optional[str] = None):
+        dvcfile, name = parse_name(name, dvcfile)
+        # this doesn't update it "in place": self.read() won't return the updated value
```
[the remark this answers is not in the document]
Comment (aguschin) [reply dvc/repo/artifacts.py:69]: Not sure I can make it work easily: to read it, you need to re-create `self.repo` that is passed upon `Artifacts` class initialization.

[the remark this answers is not in the document]
Comment (aguschin) [reply dvc/repo/artifacts.py:69]: Please share an idea how to deal with that if you have it, but I don't think it's very important right now. The only writer will be Studio, and in Studio writing is separated from reading. So it shouldn't be an issue for now.

**Comment by aguschin** [comment]
@omesser @skshetry may I ask you for a review? Thanks!

File: dvc/repo/artifacts.py [hunk 9]
```diff
@@ -50,18 +59,31 @@ def read(self) -> Dict[str, Dict[str, Artifact]]:
...
                 check_name_format(name)
                 artifacts[dvcyaml][name] = Artifact(**value)
         return artifacts
+
+    def add(self, name: str, artifact: Artifact, dvcfile: Optional[str] = None):
+        # this doesn't update it "in place": self.read() won't return the updated value
+        if not dvcfile or dvcfile[-len("dvc.yaml") :] != "dvc.yaml":
+            dvcfile = os.path.join(dvcfile or "", "dvc.yaml")
+        dvcfile_abspath = os.path.join(self.repo.root_dir, dvcfile)
+        Path(os.path.dirname(dvcfile_abspath)).mkdir(exist_ok=True, parents=True)
```
Comment (omesser) [inline dvc/repo/artifacts.py:74]: I would move this into `_update_project_file`

**Review by omesser — CHANGES_REQUESTED** [review]

File: dvc/annotations.py [hunk 10]
```diff
@@ -28,9 +28,27 @@ def to_dict(self) -> Dict[str, str]:
 
 
 @dataclass
-class Artifact(Annotation):
+class Artifact:
```
[the remark this answers is not in the document]
Comment (aguschin) [reply dvc/annotations.py:31]: `path` is required, thus it should come before all other arguments. Thus simply inheriting doesn't work - it throws errors at running. I didn't find a better solution than simply copy-pasting this code unfortunately.

File: dvc/annotations.py [hunk 11]
```diff
@@ -28,9 +28,27 @@ def to_dict(self) -> Dict[str, str]:
...
+    meta: Dict[str, Any] = field(default_factory=dict)
+
+    def update(self, **kwargs) -> "Artifact":
+        for attr, value in kwargs.items():
+            if value and hasattr(self, attr):
+                setattr(self, attr, value)
+        return self
+
+    def to_dict(self) -> Dict[str, str]:
+        return compact(asdict(self))
```
Comment (skshetry) [inline dvc/annotations.py:51]: Are you using `update()` or `add()` API?

File: dvc/repo/artifacts.py (hunk 9 above)
Comment (skshetry) [reply dvc/repo/artifacts.py:74]: We should not create directories, it's caller's responsibility.

File: dvc/repo/artifacts.py [hunk 12]
```diff
@@ -50,18 +59,31 @@ def read(self) -> Dict[str, Dict[str, Artifact]]:
...
             for name, value in dvcfile_artifacts.items():
                 check_name_format(name)
                 artifacts[dvcyaml][name] = Artifact(**value)
         return artifacts
+
+    def add(self, name: str, artifact: Artifact, dvcfile: Optional[str] = None):
+        # this doesn't update it "in place": self.read() won't return the updated value
+        if not dvcfile or dvcfile[-len("dvc.yaml") :] != "dvc.yaml":
+            dvcfile = os.path.join(dvcfile or "", "dvc.yaml")
+        dvcfile_abspath = os.path.join(self.repo.root_dir, dvcfile)
```
Comment (skshetry) [inline dvc/repo/artifacts.py:73]: no need to make an abspath

File: dvc/repo/artifacts.py [hunk 13]
```diff
@@ -50,18 +59,31 @@ def read(self) -> Dict[str, Dict[str, Artifact]]:
...
+        Path(os.path.dirname(dvcfile_abspath)).mkdir(exist_ok=True, parents=True)
+        _update_project_file(
+            ProjectFile(self.repo, dvcfile_abspath), name, artifact.to_dict()
+        )
```
Comment (skshetry) [inline dvc/repo/artifacts.py:77]: Better to keep it simple here. Also no need to create a `ProjectFile`.
```python
with modify_yaml(path) as data:
    artifacts = data.setdefault("artifacts", {})
    artifacts.update({name: artifact})
```

Comment (skshetry) [reply dvc/repo/artifacts.py:77]: I think this is more or less what you need here:

```python
with modify_yaml(dvcfile or PROJECT_FILE) as data:
    artifacts = data.setdefault("artifacts", {})
    artifacts.update({name: artifact.to_dict()})
```

File: dvc/repo/artifacts.py [hunk 14]
```diff
@@ -50,18 +59,31 @@ def read(self) -> Dict[str, Dict[str, Artifact]]:
...
+        return dvcfile_abspath
```
[the remark this answers is not in the document]
Comment (aguschin) [reply dvc/repo/artifacts.py:79]: This is for Studio. Once you add an artifact, you need to git add & commit changes. Since DVC should know where this is going to be written, I think it's valid to return that path. Can return both `Artifact` and path. WDYT @omesser?

[the remark this answers is not in the document]
Comment (aguschin) [reply dvc/repo/artifacts.py:79]: Alternative way is to calculate this on Studio side. Since we pass `dvc.yaml` path there, I can simply use it to stage changes. It can make more sense from DVC perspective, so that can be a better way. Didn't implement it first since I thought: if `dvc.yaml` path calculation changes, we're going to need to update Studio BE to address that, so wanted to keep responsibility on DVC side here.

File: dvc/annotations.py (hunk 11 above)
Comment (aguschin) [reply dvc/annotations.py:51]: No. Deleting it, thanks.

**Comment by aguschin** [comment]
Updated to your feedback @omesser @skshetry. Looking forward to get another round here and get this merged! 🎉

File: dvc/repo/artifacts.py [hunk 15]
```diff
@@ -1,43 +1,57 @@
...
+NAME_RE = re.compile(f"^{NAME}$")
+FULLNAME = f"((?P<dirname>{DIRNAME}){SEPARATOR_IN_NAME})?(?P<name>{NAME})"
+FULLNAME_RE = re.compile(f"^{FULLNAME}$")
 
 
 def name_is_compatible(name: str) -> bool:
     return bool(NAME_RE.search(name))
 
 
+wrong_name_message = (
```
Comment (skshetry) [inline dvc/repo/artifacts.py:30]: Constants should be UPPERCASED.
```suggestion
WRONG_NAME_MESSAGE = (
```

File: dvc/repo/artifacts.py [hunk 16]
```diff
@@ -1,43 +1,57 @@
...
+    "Can't use '%s' as artifact name (ID)."
```
Comment (skshetry) [inline dvc/repo/artifacts.py:31]: Use format strings where possible.
```suggestion
    "Can't use {0!r} as artifact name (ID)."
```
And below, use `wrong_name_message.format(name)

File: dvc/repo/artifacts.py [hunk 17]
```diff
@@ -50,18 +64,29 @@ def read(self) -> Dict[str, Dict[str, Artifact]]:
...
             for name, value in dvcfile_artifacts.items():
                 check_name_format(name)
                 artifacts[dvcyaml][name] = Artifact(**value)
         return artifacts
+
+    def add(self, name: str, artifact: Artifact, dvcfile: Optional[str] = None):
+        # this doesn't update it "in place": self.read() won't return the updated value
+        with self.repo.scm_context(quiet=True):
+            if not name_is_compatible(name):
+                raise InvalidArgumentError(wrong_name_message % name)
```
Comment (skshetry) [inline dvc/repo/artifacts.py:78]: I'll suggest changing `check_name_format()` to raise an exception, and use it instead. 

You can catch that exception in `Artifacts.read()` and warn and suppress.

File: dvc/repo/artifacts.py [hunk 18]
```diff
@@ -1,43 +1,57 @@
...
 
-class ArtifactsFile(FileMixin):
-    from dvc.schema import SINGLE_ARTIFACT_SCHEMA as SCHEMA
-
-    def dump(self, stage, **kwargs):
-        raise NotImplementedError
-
-    def merge(self, ancestor, other, allowed=None):
-        raise NotImplementedError
+def check_for_nested_dvc_repo(dvcfile: Path):
```
Comment (skshetry) [inline dvc/repo/artifacts.py:45]: What's the harm in allowing this?

Comment (aguschin) [reply dvc/repo/artifacts.py:45]: It's confusing, since if you add it like this, you later won't find it (it's out of scope for this DVC repo). Since Studio is the only user for this API, for now if the user adds the model like this (if we disable check), he won't find it in Studio MR. Makes sense to make this check on DVC side to me.

File: dvc/repo/artifacts.py (hunk 17 above)
Comment (aguschin) [reply dvc/repo/artifacts.py:78]: Thanks @skshetry. Appreciate your review, made the changes.

Comment (skshetry) [reply dvc/repo/artifacts.py:78]: Did you forget to push?

Comment (aguschin) [reply dvc/repo/artifacts.py]: pushed it to another branch first 😅 pushed it here now!

File: dvc/repo/artifacts.py (hunk 9 above)
Comment (omesser) [reply dvc/repo/artifacts.py]: @skshetry  - can you explain please? It's not clear to me what you're referring to . 
is "caller" being whoever calls/uses `Repo`? Or `Repo` expected to do this but not `Artifact` object (and friendts)? Is there some agreement that specific levels of the code (or API) have / don't have FS side effects?
Probably because I'm unfamiliar but the comment didn't help me understand better 😄

File: dvc/annotations.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (omesser) [reply dvc/annotations.py:31]: I don't know if there's a nicer solution here.
@skshetry  - maybe you have an idea?

**Review by omesser — APPROVED** [review]

File: dvc/repo/get.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aguschin) [reply dvc/repo/get.py]: Since we removed `get` implementation from this PR, this is irrelevant here now. WE'll get back to it once we start implementing `get`.

File: dvc/repo/artifacts.py (hunk 9 above)
Comment (skshetry) [reply dvc/repo/artifacts.py]: If you pass an invalid path, it should fail, not create it. The API should do the least that's expected from it to work to have less surprises. 

Also, the API is not designed to create artifacts in random dvc.yaml file. Most of the time, it's expected to be in the root of the dvc repo or in your current working directory. 

I guess it's following POSIX-y design rule: do one thing and do it well, in the same way `cp`/`mv` don't create directories.

Comment (omesser) [reply dvc/repo/artifacts.py]: oh, you're not talking about "code" callers in DVC, but DVC api as a whole, ok

Comment (omesser) [reply dvc/repo/artifacts.py]: I got some questions about high level behavior that may relate to those type of choices (adding a model to a non dvc initialized repo?) but it's out of scope for this comment thread I believe

File: dvc/repo/artifacts.py [hunk 19]
```diff
@@ -50,18 +58,31 @@ def read(self) -> Dict[str, Dict[str, Artifact]]:
...
-                check_name_format(name)
+                try:
+                    check_name_format(name)
+                except InvalidArgumentError as e:
+                    logger.warning(e.msg)
                 artifacts[dvcyaml][name] = Artifact(**value)
         return artifacts
+
+    def add(self, name: str, artifact: Artifact, dvcfile: Optional[str] = None):
+        # this doesn't update it "in place": self.read() won't return the updated value
```
Comment (skshetry) [inline dvc/repo/artifacts.py:72]: Remove this comment please, it’s unnecessary.

File: dvc/repo/artifacts.py (hunk 18 above)
Comment (skshetry) [reply dvc/repo/artifacts.py:39]: As i have said before, dvc does not support nedted subrepos, so it does not make sense to add an error message about this.
